### PR TITLE
Fix: do not override invalid domains

### PR DESCRIPTION
**Pull request by Loyalsoldier**
Patch borrows from https://github.com/v2ray/v2ray-core/issues/2615#issuecomment-657243545

Related issues:

https://github.com/v2ray/v2ray-core/issues/1913
https://github.com/v2ray/v2ray-core/issues/2615

### CODE REVIEW

**Comment by Loyalsoldier** [comment]
Not sure whether it will work on FakeDNS.

**Comment by rurirei** [comment]
surprisely why those invalid domains could be registered and resolved?

fakedns will ok if those could be worked resolved.

**Comment by changyp6** [comment]
Thanks for promoting my patch for this issue.

Loyalsoldier ***@***.***> 于2021年3月21日周日 上午11:27写道：

File: app/dispatcher/default.go [hunk 1]
```diff
@@ -218,8 +218,13 @@ func (d *DefaultDispatcher) Dispatch(ctx context.Context, destination net.Destin
 			if shouldOverride(result, sniffingRequest.OverrideDestinationForProtocol) {
 				domain := result.Domain()
 				newError("sniffed domain: ", domain).WriteToLog(session.ExportIDToError(ctx))
-				destination.Address = net.ParseAddress(domain)
-				ob.Target = destination
+				if strings.Contains(domain, " ") {
```
Comment (kslr) [inline app/dispatcher/default.go:221]: Is space a bad idea ?

Comment (Loyalsoldier) [reply app/dispatcher/default.go]: 如果嗅探到的域名是无效域名，则不改写目标地址为该无效域名。

**Comment by xiaokangwang** [comment]
I have to say that a better way to do this is to use configure file to do this. Since some domain name are valid but do not have a do in that, such as com. We shouldn't hard code this logic.

**Comment by rockxsj** [comment]
> I have to say that a better way to do this is to use configure file to do this. Since some domain name are valid but do not have a dot in that, such as com. We shouldn't hard code this logic.

Hi,so how to config to support domain like Mijia Cloud?

**Comment by Vigilans** [comment]
In `common/strmatcher`, we have a method `ToDomain(pattern string) (string, error)` now, which will convert a string to a valid string in domain charset (the LDH subset):
https://github.com/v2fly/v2ray-core/blob/5604863859eab4b3c4dbf84c7231bd7544b07256/common/strmatcher/matchers.go#L149-L177

It will return error if the input string cannot be converted to a domain string (e.g. containing a space). This method does not support converting UTF-8 string to puny code yet, but `nslookup <a domain containing non-ASCII string>` will automatically convert the string to puny code format. If needed we can add puny code support in the future.

By using this method, we may justify the input string passed to DNS app at the point of domain overriding and DNS outbound, where:
1. Invalid domain string will not override the destination.
2. DNS outbound will directly return empty record for invalid domain string.

If applied, bugs mentioned in https://github.com/v2fly/v2ray-core/pull/1988 would not occur even without linked PR.

**Pull request closed without merging**